lua/lspconfig/server_configurations/erlangls.lua

```diff
@@ -1,8 +1,13 @@
 local util = require 'lspconfig.util'
 
+local cmd = 'erlang_ls'
+if vim.fn.has 'win32' == 1 then
+  cmd = { 'cmd.exe', '/C', 'erlang_ls.cmd' }
```

Comment:
can't this just be erlang_ls if you are spawning via cmd?

Reply:
No. When you build the `erlang-ls` project the language server is turned into an `escript`. On unix systems, it is possible to run escripts via the shebang header:

```
$ head -n20 erlang-ls/_build/default/bin/erlang_ls
#!/usr/bin/env escript
%%
%%! -connect_all false
BINARY DATA STARTS HERE
```

Windows of course doesn't have that ability. erlang-ls ships with a `.cmd` file that does the right thing:

```
C:\Users\bakkenl\erlang-ls [main ≡]> cat .\_build\default\bin\erlang_ls.cmd
@echo off
escript.exe "%~dpn0" %*
```

My guess is this is a frequent pattern for other language servers on Windows.

Reply:
I originally had `erlang_ls.cmd` in my `PATH`, but Neovim couldn't spawn the `erlang_ls` process with an error message of "not in PATH". That's what prompted this PR. If a command does not end in an extension in `PATHEXT` Windows doesn't know what to do with it. C'est la vie.

```
C:\Users\bakkenl> echo $env:PATHEXT
.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC;.PY;.PYW;.CPL
```

Reply:
The other option would be to start `escript.exe` but then you'd also have to know the path to the generated `erlang_ls` escript file. When you build `erlang-ls` on a Windows system, the build tool (`rebar3`) is smart enough to generate the required `.cmd` file along with the generated escript file.

Reply:
Sorry, I haven't been 100% clear. On Windows, there are _two_ files that are required for the Erlang language server. The generated escript file as well as the `.cmd` to run it.

Reply:
> Right, but I thought that if we spawn via cmd.exe, we don't need to postfix with .cmd, this was the case with node/npm

I decided to install the typescript language server to see what's up, since I can see [here](https://github.com/neovim/nvim-lspconfig/blob/master/lua/lspconfig/server_configurations/tsserver.lua#L3-L8) that the `.cmd` extension is not used.

The non-`.cmd` file that is installed is a shell script, so clearly that's not the file that is executed by nvim:

```
C:\Users\bakkenl>type C:\Users\bakkenl\AppData\Roaming\npm\typescript-language-server
#!/bin/sh
basedir=$(dirname "$(echo "$0" | sed -e 's,\\,/,g')")
...
...
...
```

The other file is a real `.cmd` file:

```
C:\Users\bakkenl>type C:\Users\bakkenl\AppData\Roaming\npm\typescript-language-server.cmd
@ECHO off
GOTO start
:find_dp0
SET dp0=%~dp0
EXIT /b
:start
SETLOCAL
CALL :find_dp0

IF EXIST "%dp0%\node.exe" (
  SET "_prog=%dp0%\node.exe"
) ELSE (
  SET "_prog=node"
  SET PATHEXT=%PATHEXT:;.JS;=;%
)

endLocal & goto #_undefined_# 2>NUL || title %COMSPEC% & "%_prog%"  "%dp0%\node_modules\typescript-language-server\lib\cli.js" %*
```

So, even though you're not specifying the extension in `tsserver.lua` the `cmd.exe` process is using the `PATHEXT` environment variable to figure out what to actually run. In this case it has to be the `.cmd` batch file. So, may as well specify it here:

https://github.com/neovim/nvim-lspconfig/blob/master/lua/lspconfig/server_configurations/tsserver.lua#L7

At any rate, this is tangential to Erlang LS since using `erlang_ls.cmd` is the only way it'll work.

Reply:
In that case you can remove the cmd.exe workaround and revert to your original commit. We originally did map all node servers to the `.cmd` postfix'd command, but this introduced issues with volta (for which the non-postfixed command is correct). I am not a windows user, so I wanted something simple (as in, the same command would work for npm installed servers, volta servers, and things like erlang) without adding complexity

Reply:
> In that case you can remove the cmd.exe workaround and revert to your original commit

During our discussion I decided to do some digging into Neovim and it's use of `libuv`. It turns out my original commit worked due to [some undocumented behavior of `CreateProcessW`](https://stackoverflow.com/q/21553379/1466825). That's the win32 API function that `libuv`'s `uv_spawn` function uses under the hood.

[The `CreateProcessW` docs](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessw) state that you have to run batch files via `cmd.exe /C`:

```
To run a batch file, you must start the command interpreter; set lpApplicationName to cmd.exe and set lpCommandLine to the following arguments: /c plus the name of the batch file.

[in, out, optional] lpCommandLine
```

So, as it stands, the current commit is correct. If we want to get pendantic, the other language servers that ship `.cmd` or `.bat` files on Windows should be modified as well. I can do that in another PR.

Reply:
> If we want to get pendantic, the other language servers that ship .cmd or .bat files on Windows should be modified as well. I can do that in another PR.

They seem to work as is, can you explain the windows behavior to me? I'm fine merging this as-is then and continuing the discussion in a follow-up PR.

Reply:
> They seem to work as is, can you explain the windows behavior to me?

I've been pretty verbose today (thanks for putting up with it) and maybe you missed this comment?

https://github.com/neovim/nvim-lspconfig/pull/1566#discussion_r768113851

Reply:
> If we specify it, it will break volta and other NPM package managers

That's interesting, I'll check out volta one of these days. The two npm language servers I installed both install `.cmd` shims. But there's a chance a `.bat` or even an `.exe` would be installed, in which case leaving the extension off makes sense.

```
C:\windows\system32>npm install -g bash-language-server
npm WARN deprecated left-pad@1.3.0: use String.prototype.padStart()
npm WARN deprecated har-validator@5.1.5: this library is no longer supported
npm WARN deprecated request-promise-native@1.0.9: request-promise-native has been deprecated because it extends the now deprecated request package, see https://github.com/request/request/issues/3142
npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142

added 115 packages, and audited 116 packages in 6s

3 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities

C:\windows\system32>where bash-language-server
C:\Users\bakkenl\AppData\Roaming\npm\bash-language-server
C:\Users\bakkenl\AppData\Roaming\npm\bash-language-server.cmd
```